Added an SPDL-based DataLoader that handles concurrent IO separate from CPU processing.

alxmrs opened the pull request:
This PR is an improvement on #278 that uses learnings from a [prototype that I made in a one-off repo](https://github.com/LaureZanna/Ocean_Emulator/issues/281#issuecomment-3028790272). My bet is that this will improve data loading performance by separating the IO-bound parts of our TorchTrainDataset from the CPU-bound parts. This new loader provides an affordance for controlling the concurrency of these two aspects of the pipeline separately. Once this feature is landed, in a future task, I will work to tune these two parameters (which is dependent on the particulars of the VM in which we run the pipeline). 

Furthermore, this improves on #278 in that I got the memory model right s.t. it doesn't make a zombie thread that can't be killed. CI in this PR completes in a reasonable time frame, where as the former PR hangs indefinitely.

alxmrs committed:
Create one ThreadPoolExecutor per training run.

This should have better performance than spinup up and shutting down a thread pool per ever loop per every example.

alxmrs committed:
Made a few fixes.

alxmrs committed:
Added a new SPDL loader that uses seperate concurrency for data loading and data processing (convering to tensors).

alxmrs committed:
Fixed outstanding type issues. To do this, I made the SPDL loader a subclass of DataLoader.

alxmrs committed:
More succinct return

alxmrs committed:
Added a helpful note.

alxmrs commented on src/ocean_emulators/datasets.py:
```diff
@@ -744,18 +776,131 @@ def _get_x_index(self, idx: int, step: int) -> xr.DataArray:
+                executor=self.executor,
+                output_order="input",
+            )
+            .aggregate(self.steps)
```
IIUC, This should download all the steps concurrently. Since I pass in the same executor that is passed in, I also think a shared threadpool will be used to download the components of each Variable in these steps.

alxmrs commented on tests/test_datasets.py:
```diff
@@ -119,12 +121,26 @@ def make_loader(
-            collate_fn=collate_fn,
-        )
+        if use_spdl_loader:
+            assert version == LoaderVersion.OM4_TORCH, (
+                "SPDL Loader only works for the Torch Dataset!"
+            )
+            assert isinstance(data, ConcatDataset), "Must not be an inference dataset!"
```
These asserts aren't really necessary, but I find that they make the type checker happy.

alxmrs committed:
Fix problem with merge.

alxmrs committed:
Added back in concurrency control (merge issue).

alxmrs commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,120 @@ def load_variable_data(var: xr.Variable) -> None:
+WithSteps = tuple[int, T]
+
+
+class SpdlTorchDataLoader(DataLoader):
```
TODO(alxmrs): Add docstring to class.

alxmrs committed:
Improve documentation of the new SPDL loader.

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -638,50 +651,66 @@ def __init__(
-    def __getitem__(self, idx: int):
-        start_time = time.perf_counter()
-        TD = TrainData(self.num_prognostic_channels)
+    def perform_io(self, x_index, *, executor=None) -> tuple[xr.Dataset, xr.Dataset]:
```
How about `load_prognostic_and_boundary`?

jder replied:
Also, looks like this is always passed `self.executor`? So might as well use that directly rather than passing it?

alxmrs replied:
Both SGTM.

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -638,50 +651,66 @@ def __init__(
-                    prognostic_selected.to_array()
-                    .transpose("time", "variable", "lat", "lon")
-                    .to_numpy()
+    def prognostic_to_tensor(self, prognostic_selected):
```
Type annotations, please (and for boundary_to_tensor)

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -638,50 +651,66 @@ def __init__(
         return TD
 
-    def _get_input_and_label(
+    def join_and_partition_by_time(
```
🐑 I prefer the old name, which is about the goal/interface of the method rather than the implementation/details of it.

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+
+    def __init__(
+        self,
+        dataset: ConcatDataset,
```
It seems surprising that we are depending on using a `ConcatDataset` and ignoring all datasets in that `ConcatDataset` other than the first. At a minimum we should assert there is only one dataset (perhaps in the caller and take a TorchTrainDataset here instead?). Between this and duplicate sampler stuff it kinda smells like we have the wrong abstraction somewhere, WDYT?

alxmrs replied:
I've address this, too, as it was related to the pipeline correctness. 

TODO(alex): Address the sampler interface -- you point out a good alternative.

alxmrs replied:
wrt the sampler -- an alternative interface is an interesting consideration, esp given that I'm still not sure the best way to implement `drop_last`.

jder commented on src/ocean_emulators/train.py:
```diff
@@ -794,27 +797,59 @@ def init_data_loaders(self, cur_step: int) -> None:
+                RandomSampler
+                if self.distributed is None
+                else functools.partial(DistributedSampler, shuffle=False)
+            )
```
It puts us in an awkward state to make additional samplers here which are not the same as `self.train_sampler` / `self.val_sampler` (which are still created above) given the latter are used in `self.run`, even when `use_spdl` is true.

jder replied:
(And I continue to find functools.partial more subtle and less type-safe than eg `lambda dataset: DistributedSampler(dataset, shuffle=False)`)

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+        cpu_workers: int,
+        batch_size: int,
+        drop_last: bool = False,
+        sampler: type[Sampler] | partial[DistributedSampler] | None = None,
```
`Callable[[list[T]], Sampler[T]]` or something might be a nicer annotation here?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+        self.cpu_workers = cpu_workers
+        self.timeout = timeout
+        if pin_memory:
+            warnings.warn(f"Pinning memory is currently not supported in {self!r}.")
```
🐑 I think this should probably be an error?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+        if self._sampler_klass is not None:
+            samples = self._sampler_klass(items)
+        if self.drop_last:
+            samples = itertools.islice(samples, len(items) - 1)
```
Should this only happen if the last batch is incomplete?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+
+        # Ensure that the pipeline-level aggregation respects the steps of train data!
+        steps = [step for step, _ in downloaded]
+        assert steps == list(range(self.steps))
```
Why are we threading the steps through all these stages rather than having a `self.steps()` or whatever that each stage can access?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+        """Wire together steps into an SPDL pipeline."""
+        total_workers = self.io_workers + self.cpu_workers + 2
+        # Below, we ignore type information because this pipeline can't infer that the
+        # return type is `TrainData`.
```
Yeah, the type annotations for PipelineBuilder are pretty unhelpful. Can you fix with eg `PipelineBuilder[int, TrainData]()….pipe[int, WithSteps[xr.DataArray](…)`?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+            .pipe(self.collate_fn, concurrency=self.cpu_workers)
+            # Later, we can add a step to efficiently dispatch data to GPU.
+            .add_sink(6)
+            .build(num_threads=total_workers)
```
I don't really understand the various magic numbers in this function, can you leave an explanation? Why + 2? (Why `num_workers + 2` in the caller?) Why do we want as many threads as we have concurrency when io_workers is from a separate thread pool?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+WithSteps = tuple[int, T]
+
+
+class SpdlTorchDataLoader(DataLoader):
```
Why is this class a torch `DataLoader`? It doesn't seem like it really meets the contract of that class (eg the `__iter__` return type, unsupported `__init__` arguments). It also doesn't seem to use anything from the superclass?

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -759,3 +788,175 @@ def load_variable_data(var: xr.Variable) -> None:
+        pipeline = self._build_pipeline()
+
+        # TIL that the __exit__ condition will still be called even though we use
+        # a `return` instead of a `yield`.
```
I was very suspicious of this and so dug in a little. It's true that `__exit__` is called but it happens at the time of the `return` -- ie the pipeline is stopped before the iterator is ever iterated. Turns out this was passing the test because the pipeline silently (!) makes the iterator empty after the pipeline is stopped and `assert_equal_samples` uses `zip` without `strict=True` so `[]` will compare as equal to anything.

alxmrs replied:
!!

alxmrs replied:
Ah, foiled by Google's AI Overviews!

jder reviewed (changes requested):
Some feedback from me. My overall impression is that this doesn't seem to fit the existing abstractions very well and so left some suggestions about how to resolve. I am still worried about the immaturity of the library given the deadlocks you were getting and the pipeline typing stuff. Overall I don't think this is something we should merge without good evidence this is the a win, performance-wise so I vote we get that evidence first before resolving these comments.

alxmrs committed:
Fix to `assert_equal_samples` to not silently pass!

zip should be strict. Thanks @jder.

alxmrs committed:
Passing more tests!

Added a new protocol for default and custom loaders. Also, added correct "drop_last" implementation.

alxmrs committed:
Drop last implementation was not good after all. Removing for now.

alxmrs committed:
Fix: need to preserve the input order at every stage of the pipeline.

alxmrs committed:
A few renaming / interface changes based on the PR.

alxmrs committed:
SPDL uses multi-threaded data parallelism. Nudging to use a different backend.

alxmrs committed:
Merge branch 'main' into spdl-concurrency-alt

alxmrs commented:
Closing for now; this experiment may prove promising in the end, however, it requires changing our DDP training scheme to fully work, which is a bigger refactor than we originally bargained for.
